lib/services/role.go

```diff
@@ -118,41 +132,330 @@ type Access interface {
+				Verbs: RO(),
+			},
+		}
+	}
+
+	// restrict wildcards
+	for _, login := range r.Spec.Allow.Logins {
+		if login == Wildcard {
+			return trace.BadParameter("wildcard matcher is not allowed in logins")
+		}
+		if !cstrings.IsValidUnixUser(login) {
+			return trace.BadParameter("%q is not a valid user name", login)
+		}
+	}
+	for key, val := range r.Spec.Allow.NodeLabels {
+		if key == Wildcard && val != Wildcard {
+			return trace.BadParameter("selector *:<val> is not supported")
+		}
+	}
+
+	return nil
+}
+
+// String returns the human readable representation of a role.
+func (r *RoleV3) String() string {
+	return fmt.Sprintf("Role(Name=%v,MaxSessionTTL=%v,Options=%v,Allow=%v,Deny=%v)",
+		r.GetName(), r.GetMaxSessionTTL(), r.Spec.Options, r.Spec.Allow, r.Spec.Deny)
+}
+
+// RoleSpecV3 is role specification for RoleV3.
+type RoleSpecV3 struct {
+	// MaxSessionTTL is a maximum duration for a SSH or Web session.
+	MaxSessionTTL Duration `json:"max_session_ttl" yaml:"max_session_ttl"`
+	// Options is for OpenSSH boolean options like agent forwarding.
+	Options map[string]bool `json:"options" yaml:"options"`
+	// Allow is the set of conditions evaluated to grant access.
+	Allow RoleConditions `json:"allow,omitempty" yaml:"allow,omitempty"`
+	// Deny is the set of conditions evaluated to deny access. Deny takes priority over allow.
+	Deny RoleConditions `json:"deny,omitempty" yaml:"deny,omitempty"`
+}
+
+// RoleConditions is a set of conditions that must all match to be allowed or
+// denied access.
+type RoleConditions struct {
+	// Logins is a list of *nix system logins.
+	Logins []string `json:"logins" yaml:"logins"`
+	// Namespaces is a list of namespaces (used to partition a cluster).
+	Namespaces []string `json:"namespaces,omitempty" yaml:"namespaces,omitempty"`
+	// NodeLabels is a map of node labels (used to dynamically grant access to nodes).
+	NodeLabels map[string]string `json:"node_labels,omitempty" yaml:"node_labels,omitempty"`
+	// Rules is a list of resources to operate on.
+	Rules []RoleRule `json:"rules" yaml:"rules"`
+}
+
+// Equals returns true if the role conditions are equal and false if they are not.
+func (r *RoleConditions) Equals(o RoleConditions) bool {
+	if utils.StringSlicesEqual(r.Logins, o.Logins) == false {
+		return false
+	}
+	if utils.StringSlicesEqual(r.Namespaces, o.Namespaces) == false {
+		return false
+	}
+	if utils.StringMapsEqual(r.NodeLabels, o.NodeLabels) == false {
+		return false
+	}
+	if len(r.Rules) != len(o.Rules) {
+		return false
+	}
+	for i := 0; i < len(r.Rules); i++ {
+		if r.Rules[i].Equals(o.Rules[i]) == false {
+			return false
+		}
+	}
+	return true
+}
+
+// RoleRule is a list of resources and verbs (actions upon them).
+type RoleRule struct {
+	Resources []string `json:"resources" yaml:"resources"`
+	Verbs     []string `json:"verbs" yaml:"verbs"`
+}
+
+// Equals returns true if the role rules are equal and false if they are not.
+func (r *RoleRule) Equals(o RoleRule) bool {
```

Comment:
I would add a convenient `.Clone()` method and used it when Roles are manipulated.